core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -301,6 +303,113 @@ class TransactionStateManager(brokerId: Int,
+          s"due to ${status.error.exceptionName}")
+
+        // transform the log append error code to the corresponding coordinator error code
+        status.error match {
+          case Errors.UNKNOWN_TOPIC_OR_PARTITION
+               | Errors.NOT_ENOUGH_REPLICAS
+               | Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND
+               | Errors.REQUEST_TIMED_OUT => // note that for timed out request we return NOT_AVAILABLE error code to let client retry
+
+            debug(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning ${Errors.COORDINATOR_NOT_AVAILABLE} to the client")
+
+            Errors.COORDINATOR_NOT_AVAILABLE
+
+          case Errors.NOT_LEADER_FOR_PARTITION =>
+
+            debug(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning ${Errors.NOT_COORDINATOR} to the client")
+
+            Errors.NOT_COORDINATOR
+
+          case Errors.MESSAGE_TOO_LARGE
+               | Errors.RECORD_LIST_TOO_LARGE
+               | Errors.INVALID_FETCH_SIZE =>
+
+            error(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning UNKNOWN error code to the client")
+
+            Errors.UNKNOWN
+
+          case other =>
+            error(s"Appending metadata message $txnMetadata for $transactionalId failed due to " +
+              s"unexpected error: ${status.error.message}")
+
+            other
+        }
+      }
+
+      if (responseError != Errors.NONE) {
+        debug(s"Updating $transactionalId's transaction state to $txnMetadata for $transactionalId failed after the transaction message " +
+          s"has been appended to the log since the metadata does not match anymore.")
+      } else {
+        // now try to update the cache: we need to update the status in-place instead of
+        // overwriting the whole object to ensure synchronization
+        getTransaction(transactionalId) match {
+          case Some(metadata) =>
+            metadata synchronized {
+              if (metadata.pid == txnMetadata.pid &&
+                metadata.epoch == txnMetadata.epoch &&
+                metadata.txnTimeoutMs == txnMetadata.txnTimeoutMs &&
+                metadata.completeTransitionTo(txnMetadata.state)) {
+                // only topic-partition lists could possibly change (state should have transited in the above condition)
+                metadata.addPartitions(txnMetadata.topicPartitions.toSet)
+              } else {
+                throw new IllegalStateException(s"Completing transaction state transition to $txnMetadata while its current state is $metadata.")
+              }
+            }
+
+          case None =>
+            // this transactional id no longer exists, maybe the corresponding partition has already been migrated out.
```

Comment:
It's easy to check for this case, right? If it hasn't actually happened we can raise `IllegalStateException`?